ports/esp32/machine_can.c

```diff
@@ -0,0 +1,672 @@
+        { MP_QSTR_rtr, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_bool = false} },
+    };
+
+    // parse args
+    machine_can_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    size_t len;
+    mp_obj_t *params;
+    mp_obj_get_array(args[ARG_params].u_obj, &len, &params);
+    if (len != 2) {
+        mp_raise_ValueError("params shall be a 2-values list");
+    }
+    uint32_t id = mp_obj_get_int(params[0]);
+    uint32_t mask = mp_obj_get_int(params[1]); // FIXME: Overflow in case 0xFFFFFFFF for mask
+    if (args[ARG_mode].u_int == FILTER_RAW_SINGLE || args[ARG_mode].u_int == FILTER_RAW_DUAL) {
+        self->config->filter.single_filter = (args[ARG_mode].u_int == FILTER_RAW_SINGLE);
+        self->config->filter.acceptance_code = id;
+        self->config->filter.acceptance_mask = mask;
+    } else {
+        self->config->filter.single_filter = self->extframe;
+        _machine_hw_can_set_filter(self, id, mask, args[ARG_bank].u_int, args[ARG_rtr].u_int);
+    }
+    ESP_STATUS_CHECK( can_stop() );
+    ESP_STATUS_CHECK( can_driver_uninstall() );
+    ESP_STATUS_CHECK( can_driver_install(
+                          &self->config->general,
+                          &self->config->timing,
+                          &self->config->filter) );
+    ESP_STATUS_CHECK( can_start() );
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(machine_hw_can_setfilter_obj, 1, machine_hw_can_setfilter);
+
+STATIC void machine_hw_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_can_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    if (self->config->initialized) {
+        qstr mode;
+        switch (self->config->general.mode)
+        {
+        case CAN_MODE_LISTEN_ONLY:
+            mode = MP_QSTR_LISTEN;
+            break;
+        case CAN_MODE_NO_ACK:
+            mode = MP_QSTR_NO_ACK;
+            break;
+        case CAN_MODE_NORMAL:
+            mode = MP_QSTR_NORMAL;
+            break;
+        default:
+            mode = MP_QSTR_UNKNOWN;
+            break;
+        }
+        mp_printf(print,
+                  "CAN(tx=%u, rx=%u, baudrate=%ukb, mode=%q, loopback=%u, extframe=%u)",
+                  self->config->general.tx_io,
+                  self->config->general.rx_io,
+                  self->config->baudrate,
+                  mode,
+                  self->loopback,
+                  self->extframe);
+    } else {
+        mp_printf(print, "CAN Device is not initialized");
+    }
+}
+
+// init(tx, rx, baudrate, mode = CAN_MODE_NORMAL, tx_queue = 2, rx_queue = 5)
+STATIC mp_obj_t machine_hw_can_init(size_t n_args, const mp_obj_t *args, mp_map_t *kw_args) {
+    machine_can_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    if (self->config->initialized) {
+        ESP_LOGW(DEVICE_NAME, "Device is already initialized");
```

Comment:
Why does this not raise an error? MP doesn't use ESP_LOGW in general...

Reply:
@twe @dpgeorge @jimmo 
I find it difficult to choose the type of error in cases
"Device is already initialized" and "Device is not initialized"
I can raise
_mp_raise_ValueError(msg);_
_mp_raise_TypeError(msg);_
_mp_raise_NotImplementedError(msg);_
_mp_raise_OSError(errno);_
but I don't like them.
In fact, it is an algorithm user program bug.
Twice init, twice deinit etc.

I want to help a user change his program.
I prefer to raise 
mp_raise_OSError(EPERM, "Device is already initialized")  
 with console output like:
**>>>** OSError: [Errno 1] EPERM: Device is already initialized